module/surface_sam_sid_vhf.c

```diff
@@ -0,0 +1,411 @@
+} __packed;
+
+struct vhf_meta_meta_resp {
+    u32 len;
+    u16 vendor_id;
+    u16 product_id;
+    u8  _1[24];
+} __packed;
+
+union vhf_buffer_data {
+	struct vhf_meta_info_resp info;
+	u8 pld[0x76];
+	struct vhf_meta_meta_resp meta;
+};
+
+struct surface_sam_sid_vhf_meta_resp {
+	struct surface_sam_sid_vhf_meta_rqst rqst;
+	union vhf_buffer_data data;
+} __packed;
+
+
+static int vhf_get_metadata(u8 iid, struct vhf_meta_meta_resp *meta)
+{
+	int status;
+	int len;
+	u8 *buf;
+
+	struct surface_sam_sid_vhf_meta_resp *resp;
+	struct surface_sam_ssh_rqst rqst = {};
+	struct surface_sam_ssh_buf result = {};
+
+	resp = kzalloc(sizeof(struct surface_sam_sid_vhf_meta_resp), GFP_KERNEL);
+
+	resp.rqst.id = 2;
+	resp.rqst.offset = 0;
+	resp.rqst.limit = 0x76;
+	resp.rqst.end = 0;
+
+	rqst.tc = 0x15;
+	rqst.cid = 0x04;
+	rqst.iid = iid;
+	rqst.pri = 0x02;
+	rqst.snc = 0x01;
+	rqst.cdl = sizeof(struct surface_sam_sid_vhf_meta_rqst);
+	rqst.pld = (u8*)&resp.rqst;
+
+	result.cap  = sizeof(struct surface_sam_sid_vhf_meta_resp);
+	result.len  = 0;
+	result.data = (u8*)resp;
+
+	status = surface_sam_ssh_rqst(&rqst, &result);
+	if (status) {
+		return status;
+	}
+
+	*meta = resp;
+}
+
+static int vhf_get_hid_descriptor(struct hid_device *hid, u8 iid, u8 **desc, int *size)
+{
+	int status;
+	int len;
+	u8 *buf;
+
+	struct surface_sam_ssh_rqst rqst = {};
+	struct surface_sam_sid_vhf_meta_resp resp = {};
+	struct surface_sam_ssh_buf result = {};
+
+	resp.rqst.id = 0;
+	resp.rqst.offset = 0;
+	resp.rqst.limit = 0x76;
+	resp.rqst.end = 0;
+
+	rqst.tc = 0x15;
+	rqst.cid = 0x04;
+	rqst.iid = iid;
+	rqst.pri = 0x02;
+	rqst.snc = 0x01;
+	rqst.cdl = sizeof(struct surface_sam_sid_vhf_meta_rqst);
+	rqst.pld = (u8*)&resp.rqst;
+
+	result.cap  = sizeof(struct surface_sam_sid_vhf_meta_resp);
+	result.len  = 0;
+	result.data = (u8*)&resp;
+
+	// first fetch 01 to get the total length
+	status = surface_sam_ssh_rqst(&rqst, &result);
+	if (status) {
+		return status;
+	}
+
+	len = resp.data.info.hid_len;
+
+	// allocate a buffer for the descriptor
+	buf = kzalloc(len, GFP_KERNEL);
+
+	// then, iterate and write into buffer, copying out resp.rqst.hid_len bytes
+	// (it gets set to the length of data read)
+
+	resp.rqst.id = 1;
+	resp.rqst.offset = 0;
+	resp.rqst.limit = 0x76;
+	resp.rqst.end = 0;
+
+	while (resp.rqst.end != 0x01 && resp.rqst.offset < len) {
+		status = surface_sam_ssh_rqst(&rqst, &result);
+		if (status) {
+			kfree(buf);
+			return status;
+		}
+		memcpy(buf + resp.rqst.offset, resp.data.pld, resp.rqst.limit);
+
+		resp.rqst.offset += resp.rqst.limit;
+	}
+
+	*desc = buf;
+	*size = len;
+
+	return 0;
+}
+
+static int sid_vhf_hid_parse(struct hid_device *hid)
+{
+	int ret = 0, size;
+	u8 *buf;
+
+	ret = vhf_get_hid_descriptor(hid, 0x00, &buf, &size);
+	if (ret != 0) {
+		hid_dbg(hid, "vhf_get_hid_descriptor ret %d\n", ret);
+		return -EIO;
+	}
+	print_hex_dump_debug("descriptor:", DUMP_PREFIX_OFFSET, 16, 1, buf, size, false);
```

Comment:
If we leave this in, then add a call to `hid_dbg` above explaining what the data means (like "HID descriptor of device:"). This will also give anyone reading the debug log an association what driver this belongs to (I think at the moment this will result in a more or less random seeming `descriptor: ...<some bytes>` without any description as to what this data belongs).

Reply:
Okay, I modified the message and decided to leave it in because it could be useful when diagnosing if we get github issues for other Surface devices.